crypto/hpke/hpke_util.c

```diff
@@ -12,24 +12,203 @@
 #include <openssl/params.h>
 #include <openssl/err.h>
 #include <openssl/proverr.h>
-#include "crypto/hpke.h"
+#include <openssl/hpke.h>
+#include <openssl/sha.h>
+#include <openssl/rand.h>
+#include "crypto/ecx.h"
+#include "internal/hpke_util.h"
 #include "internal/packet.h"
 
+/* max length of string we'll try map to a suite */
+#define OSSL_HPKE_MAX_SUITESTR 38
+
+/* Define HPKE labels from RFC9180 in hex for EBCDIC compatibility */
+/* ASCII: "HPKE-v1", in hex for EBCDIC compatibility */
+static const char LABEL_HPKEV1[] = "\x48\x50\x4B\x45\x2D\x76\x31";
+
 /*
- * The largest value happens inside dhkem_extract_and_expand
- * Which consists of a max dkmlen of 2*privkeylen + suiteid + small label
+ * Note that if additions are made to the set of IANA codepoints
+ * and the tables below, corresponding additions should also be
+ * made to the synonymtab tables a little further down so that
+ * OSSL_HPKE_str2suite() continues to function correctly.
+ *
+ * The canonical place to check for IANA registered codepoints
+ * is: https://www.iana.org/assignments/hpke/hpke.xhtml
  */
-#define LABELED_EXTRACT_SIZE (10 + 12 + 2 * OSSL_HPKE_MAX_PRIVATE)
 
 /*
- * The largest value happens inside dhkem_extract_and_expand
- * Which consists of a prklen of secretlen + contextlen of 3 encoded public keys
- * + suiteid + small label
+ * @brief table of KEMs
+ * See RFC9180 Section 7.1 "Table 2 KEM IDs"
  */
-#define LABELED_EXPAND_SIZE (LABELED_EXTRACT_SIZE + 3 * OSSL_HPKE_MAX_PUBLIC)
+static const OSSL_HPKE_KEM_INFO hpke_kem_tab[] = {
+    { OSSL_HPKE_KEM_ID_P256, "EC", OSSL_HPKE_KEMSTR_P256,
+      LN_sha256, SHA256_DIGEST_LENGTH, 65, 65, 32, 0xFF },
+    { OSSL_HPKE_KEM_ID_P384, "EC", OSSL_HPKE_KEMSTR_P384,
+      LN_sha384, SHA384_DIGEST_LENGTH, 97, 97, 48, 0xFF },
+    { OSSL_HPKE_KEM_ID_P521, "EC", OSSL_HPKE_KEMSTR_P521,
+      LN_sha512, SHA512_DIGEST_LENGTH, 133, 133, 66, 0x01 },
+#ifndef OPENSSL_NO_EC
```

Comment:
Can't the EC be pulled from a third party provider?

Reply:
I'm sure I'm not understanding what's right here tbh;-) Without that ``#ifndef OPENSSL_NO_EC`` some of the CI builds fail. If I move that ``#ifndef`` to cover the entire table, then I also get build failure when ``OPENSSL_NO_EC`` is defined, e.g.: with 
``` 
CC=clang ./config --banner=Configured --strict-warnings no-ec
```
and with the ``#ifndef`` covering all the table entries, I get:
```
crypto/hpke/hpke_util.c:43:50: error: use of GNU empty initializer extension [-Werror,-Wgnu-empty-initializer]
static const OSSL_HPKE_KEM_INFO hpke_kem_tab[] = {
                                                 ^
crypto/hpke/hpke_util.c:43:50: error: zero size arrays are an extension [-Werror,-Wzero-length-array]
2 errors generated.
```
I don't know what's the diff between the nist and cfrg curve implementations that causes that effect. I do agree though that yes HPKE needs some elliptic curve crypto of course.

For completeness, with the above ``CC=clang ./config --banner=Configured --strict-warnings no-ec`` and the current code, ``test/hpke_test`` fails but the overall ``make test`` works because I just don't run that test when ``no-ec`` is defined, which does feel a bit like a cheat.

So basically, guidance welcome:-)

Reply:
IMO if OpenSSL has been compiled with no-ec then the person building has accepted that OpenSSL will not support anything requiring ec support - so I would be tempted to just disable HPKE entirely in that case

Reply:
You can disable code conditionally via build.info

Reply:
I still want to play with this a bit more. (Will push changes then.)

I guess the issue is that, in future, (with PQC), it might be that one could use HPKE in a ``no-ec`` build. I'm tempted to just do as suggested (i.e. let that be figured out later) and make almost all HPKE calls fail for now in such builds, but will check it out a bit more first. (tomorrow)

Reply:
So what I'm thinking here is to create a KEM table entry with the ``OSSL_HPKE_KEM_ID_RESERVED`` KEM id when in a ``no-ec`` build and all the real ones are ``#ifdef'd`` out. That means that the code can be built and it's just a case of skipping the test. So later when/if a non-EC KEM is added to the table there'd be less change. Seems to work but might be a bit hacky for ya so comment welcome once I push.